### PR TITLE
fix: fixed issue where stars metric chart was flipping horizontally

### LINKED ISSUE #3073: Bug: The flippening 2: The return of flipper 🐬 - Repo pages "Stars per day" data flips when tabbing away
### Describe the bug

The "Stars per day" chart data flips when tabbing away. It flips from showing the _first_ day to the right to the _last_ day to the right:

the-flippening

Note how it shows March 1st on the right when it flips. And then when it goes back, it shows April 1st on the right (which should be the right Y axis value for time growing to the right).

Related:
- https://github.com/open-sauced/app/pull/3063#pullrequestreview-1972051474
- https://github.com/open-sauced/app/pull/3063#issuecomment-2030668053

### Steps to reproduce

1. Go to https://app.opensauced.pizza/s/kubernetes/kubernetes
2. Click to different tab
3. Click back
4. Notice chart flip

### Do you have any images or screen captures?

_No response_

### Browsers

Firefox

### Additional context (Is this in dev or production?)

_No response_

### Code of Conduct

- [X] I agree to follow this project's Code of Conduct

### Contributing Docs

- [X] I agree to follow this project's Contribution Docs

### PR DESCRIPTION
**Pull request by nickytonline**
## Description

Fixes the "flippening" issue that @jpmcb reported.

## Related Tickets & Documents

Fixes #3073

## Mobile & Desktop Screenshots/Recordings

**Before**

the-flippening

**After**

CleanShot 2024-04-02 at 14 34 22

## Steps to QA

1. Visit https://deploy-preview-3091--oss-insights.netlify.app/s/kubernetes/kubernetes
2. Go to another browser tab
3. Come back to the tab that has https://deploy-preview-3091--oss-insights.netlify.app/s/kubernetes/kubernetes
4. The graphs no longer flip.

## Tier (staff will fill in)

- [ ] Tier 1
- [ ] Tier 2
- [x] Tier 3
- [ ] Tier 4

### CODE REVIEW

**Review by nickytonline — COMMENTED** [review]
We should consider this eslint rule to avoid this issue in the future, https://github.com/cartant/eslint-plugin-etc/blob/main/docs/rules/no-assign-mutated-array.md

**Comment by brandonroberts** [comment]
I see a flipped chart for Starts per day when resized to mobile

image


image

**Comment by nickytonline** [comment]
> I see a flipped chart for Starts per day when resized to mobile
> 
> image image

Oh nos. Taking a peek again.

**Review by jpmcb — COMMENTED** [review]
The flipping still happens by adjusting the range:

the-flippening-2

(notice how the dates for the "Stars per day" chart flip from the right being April 2nd to the left becoming April 2nd)

**Review by zeucapua — APPROVED** [review]

**Comment by zeucapua** [comment]
I don't see the flip on Arc, but going to Safari does the flipping. Weird...

**Comment by nickytonline** [comment]
Looking at the data and what's rendered, something is off. The data is by date in ascending order, but the cart is showing it in reverse still.

CleanShot 2024-04-03 at 12 41 55

File: lib/hooks/api/useFetchMetricStats.ts [hunk 1]
```diff
@@ -13,10 +14,11 @@ export type StatsType = {
   forks_count?: number;
 };
 
-export function useFetchMetricStats({ repository, variant, range }: UseFetchMetricStatsParams) {
+export function useFetchMetricStats({ repository, variant, range, orderDirection = "ASC" }: UseFetchMetricStatsParams) {
```
Comment (nickytonline) [inline lib/hooks/api/useFetchMetricStats.ts:17]: The default order is descending in the API. I opted for going ascending here and avoiding reversing the array client-side.

File: components/Graphs/MetricCard.tsx [hunk 2]
```diff
@@ -10,38 +11,42 @@ type MetricCardProps = {
 };
 
 export default function MetricCard({ stats, variant, range }: MetricCardProps) {
-  const seriesData = stats
-    ?.map((stat) => (variant === "stars" ? stat.star_count || 0 : stat.forks_count || 0))
-    .reverse();
-  const bucketData = stats?.map((stat) => new Date(stat.bucket).toDateString()).reverse();
+  const countProperty = variant === "stars" ? "star_count" : "forks_count";
+  const seriesData = useMemo(() => stats?.map((stat) => stat[countProperty] ?? 0) ?? [], [countProperty, stats]);
```
Comment (nickytonline) [inline components/Graphs/MetricCard.tsx:15]: These are things we shouldn't have to recalculate unless the props change. `useMemo` was a good fit for this. Thanks Million Lint for the assist!

Comment (bdougie) [reply components/Graphs/MetricCard.tsx:15]: I'd love to see a Million Lint demo if you have time.

Comment (nickytonline) [reply components/Graphs/MetricCard.tsx:15]: Just got off a live steam with Toby from the team where we went through it. Happy to do a demo.

**Review by brandonroberts — APPROVED** [review]

**Review by zeucapua — APPROVED** [review]
backflip-happy-gif

File: components/Graphs/MetricCard.tsx (hunk 2 above)
Comment (nickytonline) [reply components/Graphs/MetricCard.tsx:15]: https://github.com/open-sauced/app/issues/3104

**Pull request merged**